File: replay/protocol/encoder.go
// Copyright (C) 2014 The Android Open Source Project
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package protocol

import (
	"encoding/binary"
	"io"
	"math"
)

// Encoder provides methods for encoding values to an io.Writer.
type Encoder struct {
	writer    io.Writer
	tmp       [8]byte
	byteOrder binary.ByteOrder
}

// NewEncoder creates an Encoder that writes to the supplied stream, with the
// specified endianness.
func NewEncoder(writer io.Writer, byteOrder binary.ByteOrder) *Encoder {
	return &Encoder{writer: writer, byteOrder: byteOrder}
}

// Write implements the io.Writer interface, delegating to the underlying writer.
func (e *Encoder) Write(p []byte) (int, error) {
	return e.writer.Write(p)
}

// Bool encodes a boolean value to the Encoder's io.Writer.
func (e *Encoder) Bool(v bool) error {
	if v {
		e.tmp[0] = 1
	} else {
		e.tmp[0] = 0
	}
	_, err := e.writer.Write(e.tmp[:1])
	return err
}

// Int8 encodes a signed, 8 bit integer value to the Encoder's io.Writer.
func (e *Encoder) Int8(v int8) error {
	e.tmp[0] = uint8(v)
	_, err := e.writer.Write(e.tmp[:1])
	return err
}

// Uint8 encodes an unsigned, 8 bit integer value to the Encoder's io.Writer.
func (e *Encoder) Uint8(v uint8) error {
	e.tmp[0] = v
	_, err := e.writer.Write(e.tmp[:1])
	return err
}

// Int16 encodes a signed, 16 bit integer value to the Encoder's io.Writer.
func (e *Encoder) Int16(v int16) error {
	e.byteOrder.PutUint16(e.tmp[:], uint16(v))
	_, err := e.writer.Write(e.tmp[:2])
	return err
}

// Uint16 encodes an unsigned, 16 bit integer value to the Encoder's io.Writer.
func (e *Encoder) Uint16(v uint16) error {
	e.byteOrder.PutUint16(e.tmp[:], v)
	_, err := e.writer.Write(e.tmp[:2])
	return err
}

// Int32 encodes a signed, 32 bit integer value to the Encoder's io.Writer.
func (e *Encoder) Int32(v int32) error {
	e.byteOrder.PutUint32(e.tmp[:], uint32(v))
	_, err := e.writer.Write(e.tmp[:4])
	return err
}

// Uint32 encodes an usigned, 32 bit integer value to the Encoder's io.Writer.
func (e *Encoder) Uint32(v uint32) error {
	e.byteOrder.PutUint32(e.tmp[:], v)
	_, err := e.writer.Write(e.tmp[:4])
	return err
}

// Float32 encodes a 32 bit floating-point value to the Encoder's io.Writer.
func (e *Encoder) Float32(v float32) error {
	e.byteOrder.PutUint32(e.tmp[:], math.Float32bits(v))
	_, err := e.writer.Write(e.tmp[:4])
	return err
}

// Int64 encodes a signed, 64 bit integer value to the Encoder's io.Writer.
func (e *Encoder) Int64(v int64) error {
	e.byteOrder.PutUint64(e.tmp[:], uint64(v))
	_, err := e.writer.Write(e.tmp[:8])
	return err
}

// Uint64 encodes an unsigned, 64 bit integer value to the Encoders's io.Writer.
func (e *Encoder) Uint64(v uint64) error {
	e.byteOrder.PutUint64(e.tmp[:], v)
	_, err := e.writer.Write(e.tmp[:8])
	return err
}

// Float64 encodes a 64 bit floating-point value to the Encoder's io.Writer.
func (e *Encoder) Float64(v float64) error {
	e.byteOrder.PutUint64(e.tmp[:], math.Float64bits(v))
	_, err := e.writer.Write(e.tmp[:8])
	return err
}

// String encodes a string (pascal-style) to the Encoder's io.Writer.
func (e *Encoder) String(v string) error {
	if err := e.Uint32(uint32(len(v))); err != nil {
		return err
	}
	_, err := e.writer.Write([]byte(v))
	return err
}

// String encodes a string (c-style) to the Encoder's io.Writer.
func (e *Encoder) CString(v string) error {
	if _, err := e.writer.Write([]byte(v)); err != nil {
		return err
	}
	return e.Uint8(0)
}

File: cc/gapic/encoder.cpp
/*
 * Copyright 2015, The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "encoder.h"
#include "stream_writer.h"

#include <cstring>

namespace gapic {

Encoder::Encoder(std::shared_ptr<StreamWriter> output) : mOutput(output), mLastObjectId(0) {}

void Encoder::Bool(bool v) {
    uint8_t b = v ? 1 : 0;
    mOutput->Write(&b, 1);
}

void Encoder::Int8(int8_t v) {
    mOutput->Write(&v, 1);
}

void Encoder::Uint8(uint8_t v) {
    mOutput->Write(&v, 1);
}

void Encoder::Uint16(uint16_t v) {
    uint8_t buf[9];
    uint16_t space = 0x7f;
    uint8_t tag = 0;
    for (int o = 8; true; o--) {
        if (v <= space) {
            buf[o] = uint8_t(v) | uint8_t(tag);
            mOutput->Write(buf + o, 9 - o);
            return;
        }
        buf[o] = uint8_t(v);
        v = v >> 8;
        space >>= 1;
        tag = (tag >> 1) | 0x80;
    }
}

void Encoder::Int16(int16_t v) {
    uint16_t uv = uint16_t(v) << 1;
    Uint16((v < 0) ? ~uv : uv);
}

void Encoder::Float32(float v) {
    uint32_t bits = *reinterpret_cast<uint32_t*>(&v);
    uint32_t shuffled =
            ((bits & 0x000000ff) << 24) |
            ((bits & 0x0000ff00) << 8)  |
            ((bits & 0x00ff0000) >> 8)  |
            ((bits & 0xff000000) >> 24);
    return Uint32(shuffled);
}

void Encoder::Uint32(uint32_t v) {
    uint8_t buf[9];
    uint32_t space = 0x7f;
    uint8_t tag = 0;
    for (int o = 8; true; o--) {
        if (v <= space) {
            buf[o] = uint8_t(v) | uint8_t(tag);
            mOutput->Write(buf + o, 9 - o);
            return;
        }
        buf[o] = uint8_t(v);
        v = v >> 8;
        space >>= 1;
        tag = (tag >> 1) | 0x80;
    }
}

void Encoder::Int32(int32_t v) {
    uint32_t uv = uint32_t(v) << 1;
    Uint32((v < 0) ? ~uv : uv);
}

void Encoder::Float64(double v) {
    uint64_t bits = *reinterpret_cast<uint64_t*>(&v);
    uint64_t shuffled =
            ((bits & 0x00000000000000ffULL) << 56) |
            ((bits & 0x000000000000ff00ULL) << 40) |
            ((bits & 0x0000000000ff0000ULL) << 24) |
            ((bits & 0x00000000ff000000ULL) << 8)  |
            ((bits & 0x000000ff00000000ULL) >> 8)  |
            ((bits & 0x0000ff0000000000ULL) >> 24) |
            ((bits & 0x00ff000000000000ULL) >> 40) |
            ((bits & 0xff00000000000000ULL) >> 56);
    return Uint64(shuffled);
}

void Encoder::Uint64(uint64_t v) {
    uint8_t buf[9];
    uint64_t space = 0x7f;
    uint8_t tag = 0;
    for (int o = 8; true; o--) {
        if (v <= space) {
            buf[o] = uint8_t(v) | uint8_t(tag);
            mOutput->Write(buf + o, 9 - o);
            return;
        }
        buf[o] = uint8_t(v);
        v = v >> 8;
        space >>= 1;
        tag = (tag >> 1) | 0x80;
    }
}

void Encoder::Int64(int64_t v) {
    uint64_t uv = uint64_t(v) << 1;
    Uint64((v < 0) ? ~uv : uv);
}

void Encoder::String(const char* v) {
    uint32_t len = v != nullptr ? static_cast<uint32_t>(strlen(v)) : 0;
    Uint32(len);
    mOutput->Write(v, len);
}

void Encoder::Data(const void* ptr, int32_t size) {
    Uint32(size);
    mOutput->Write(ptr, size);
}

void Encoder::Id(const gapic::Id& id) {
    auto it = mIds.find(id);
    if (it != mIds.end()) {
        Uint32(it->second << 1);
    } else {
        uint32_t sid = mIds.size() + 1;
        mIds[id] = sid;
        Uint32((sid << 1) | 1);
        mOutput->Write(&id.data, 20);
    }
}

void Encoder::Value(const Encodable* obj) {
    obj->Encode(this);
}

void Encoder::Variant(const Encodable* obj) {
    if (obj == nullptr) {
        Id(gapic::Id());
        return;
    }
    Id(obj->Id());
    Value(obj);
}

void Encoder::Object(const Encodable* obj) {
    if (obj == nullptr) {
        Uint32(0);
        return;
    }
    uint32_t sid = ++mLastObjectId;
    Uint32((sid << 1) | 1);
    Variant(obj);
}

} // namespace gapic
